src/cdk/a11y/input-modality/input-modality-detector.ts

```diff
@@ -0,0 +1,184 @@
+ * with a screen reader is akin to visually scanning a page, and should not be interpreted as actual
+ * user input interaction.
+ *
+ * When a user is not navigating but *interacting* with a screen reader, this service's behavior is
+ * largely undefined and depends on the events fired. For example, in VoiceOver, no keyboard events
+ * are fired when performing an element's default action via Caps Lock + Space, thus no input
+ * modality is detected.
+ */
+@Injectable({ providedIn: 'root' })
+export class InputModalityDetector implements OnDestroy {
+  /** Emits when the input modality changes. */
+  readonly inputModalityChange: Observable<InputModality>;
+
+  /** Returns the most recently detected input modality. */
+  get inputModality(): InputModality {
+    return this._inputModality.value;
+  }
+
+  /** The underlying BehaviorSubject that emits whenever an input modality is detected. */
+  private readonly _inputModality = new BehaviorSubject<InputModality>(null);
+
+  /** Options for this InputModalityDetector. */
+  private readonly _options: InputModalityDetectorOptions;
+
+  /**
+   * The timestamp of the last touch input modality. Used to determine whether mousedown events
+   * should be attributed to mouse or touch.
+   */
+  private _lastTouchMs = 0;
+
+  /**
+   * Handles keyboard events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onKeydown = (event: KeyboardEvent) => {
+    // If this is one of the keys we should ignore, then ignore it and don't update the input
+    // modality to keyboard.
+    if (this._options?.ignoreKeys?.some(keyCode => keyCode === event.keyCode)) { return; }
+
+    this._inputModality.next('keyboard');
+  }
+
+  /**
+   * Handles mouse events. Must be an arrow function in order to preserve the context when it gets
+   * bound.
+   */
+  private _onMousedown = (event: MouseEvent) => {
+    if (isFakeMousedownFromScreenReader(event)) { return; }
+
+    // Touches trigger both touch and mouse events, so we need to distinguish between mouse events
+    // that were triggered via mouse vs touch. To do so, check if the mouse event occurs closely
+    // after the previous touch event.
+    if (Date.now() - this._lastTouchMs < TOUCH_BUFFER_MS) { return; }
+
+    this._inputModality.next('mouse');
+  }
+
+  /**
+   * Handles touchstart events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onTouchstart = (event: TouchEvent) => {
+    if (isFakeTouchstartFromScreenReader(event)) { return; }
+
+    // Store the timestamp of this touch event, as it's used to distinguish between mouse events
+    // triggered via mouse vs touch.
+    this._lastTouchMs = Date.now();
+
+    this._inputModality.next('touch');
+  }
+
+  constructor(
+      private readonly _platform: Platform,
+      ngZone: NgZone,
+      @Inject(DOCUMENT) document: Document,
```

Comment:
```suggestion
      @Inject(DOCUMENT) document: Document|Document,
```
Using `Document` here will break in SSR because TypeScript preserves type information on decorated params, and `Document` doesn't existing in a nodejs environment. We can use this weird workaround to trick TS into not emitting the type as-is. 

Can go in a follow-up, but you can test this by adding something to make this API run as part of our `kitchen-sink.ts` component (which is our SSR test component).

Reply:
See the thread at https://github.com/angular/components/pull/22371#discussion_r605479216. The kitchen sink e2e test seems to pass - what should I be looking for?

Reply:
Hmm, maybe something changed in the tooling so that we don't have to worry about this any more ¯\\_(ツ)_/¯